typed_python/PythonSerializationContext.hpp

```diff
@@ -153,89 +236,61 @@ class PythonSerializationContext : public SerializationContext {
-                serializePythonObject(PyList_GetItem(o,k), b);
-            }
-        }
+        serializeIndexable(o, b, PyList_Size, PyList_Get_Item_No_Checks);
     }
 
     PyObject* deserializePyList(DeserializationBuffer& b) const {
-        uint32_t id = b.read_uint32();
-
-        PyObject* res = (PyObject*)b.lookupCachedPointer(id);
-        if (res) {
-            return incref(res);
-        }
-
-        size_t sz = b.read_uint32();
-
-        res = PyList_New(sz);
-        b.addCachedPointer(id, incref(res), true);
-
-        try {
-            for (long k = 0; k < sz; k++) {
-                PyList_SET_ITEM(res, k, deserializePythonObject(b));
-            }
-        } catch(...) {
-            Py_DECREF(res);
-            throw;
-        }
-
-        return res;
+        return deserializeIndexable(b, PyList_New, PyList_Set_Item_No_Checks);
     }
 
     void serializePyTuple(PyObject* o, SerializationBuffer& b) const {
-        std::pair<uint32_t, bool> idAndIsNew = b.cachePointer(o);
-        b.write_uint32(idAndIsNew.first);
-
-        if (idAndIsNew.second) {
-            size_t sz = PyTuple_Size(o);
+        serializeIndexable(o, b, PyTuple_Size, PyTuple_Get_Item_No_Checks);
+    }
 
-            b.write_uint32(sz);
+    PyObject* deserializePyTuple(DeserializationBuffer& b) const {
+        return deserializeIndexable(b, PyTuple_New, PyTuple_Set_Item_No_Checks);
+    }
 
-            for (long k = 0; k < sz; k++) {
-                serializePythonObject(PyTuple_GetItem(o,k), b);
-            }
-        }
+    void serializePySet(PyObject* o, SerializationBuffer& b) const {
+        serializeIterable(o, b, PySet_Size);
     }
-    PyObject* deserializePyTuple(DeserializationBuffer& b) const {
-        uint32_t id = b.read_uint32();
 
-        PyObject* res = (PyObject*)b.lookupCachedPointer(id);
-        if (res) {
-            return incref(res);
-        }
+    PyObject* deserializePySet(DeserializationBuffer &b) const {
+        return deserializeIterable(b, PySet_New, PySet_Add);
+    }
 
-        size_t sz = b.read_uint32();
+    void serializePyFrozenSet(PyObject* o, SerializationBuffer& b) const {
+        serializeIterable(o, b, PySet_Size);
+    }
 
-        res = PyTuple_New(sz);
-        b.addCachedPointer(id, incref(res), true);
+    PyObject* deserializePyFrozenSet(DeserializationBuffer &b) const {
+        // Looks like the only way to initialize a frozenset in the API is to give
+        // the constructor an iterable, which we can do only after we deserialize.
+        // So, we deserialize to a set, then use that to build the frozenset.
+        PyObject* set = deserializeIterable(b, PySet_New, PySet_Add);
+        if (!set) {
+            return set;
+        }
 
+        PyObject* res = nullptr;
         try {
-            for (long k = 0; k < sz; k++) {
-                PyTuple_SET_ITEM(res, k, deserializePythonObject(b));
-            }
+            res = PyFrozenSet_New(PyObject_GetIter(set));
```

Comment:
don't forget to decref the iterator, and also to check for exceptions during set construction